django/thunderstore/repository/models/package.py

```diff
@@ -272,6 +299,8 @@ def recache_latest(self):
         if hasattr(self, "available_versions"):
             del self.available_versions  # Bust the version cache
         self.latest = self.available_versions.first()
+        if self.latest is None:
+            self.latest = self.versions.filter(is_active=True).first()
```

Comment:
I'm fairly confident this will break some assumptions around the codebase elsewhere, as those assumptions generally have been along the lines of "The package set on the `latest` field is active & available for users". Conceptually it's not the latest version, but rather the latest valid version, meaning it should be empty if there are no available versions of said package.

I assume the reason you changed this is because some views elsewhere depend on it, to which I'd suggest directly modifying the views to account for the scenario where latest is empty. It might be worth renaming the field to `latest_available` too to avoid ambiguity around it in the future.

Reply:
I searched through every instance of latest and cached properties that use latest that I could find in the project and it doesn't seem like this breaks any assumptions in the codebase (and it doesn't fail any tests), so I think it's better to assume that a latest version can be unavailable rather than update the views and cached properties that it's already used in.

Reply:
If that's true, what's preventing us from always simply setting the latest to truly be the latest irrespective of the availability of it? I'm fairly certain there's a reason for it, perhaps it just happens to not cause issues _right now_ due to such a scenario not occurring naturally, but very well could in the future if we make more code changes.

Main point being, the current change will introduce more complexity to what this field represents (and when), and I'm not sure if it provides anything useful in return? Just consider the spec difference when spelled out:

Previously:
> Points to the latest public version of the package

Proposed change:
> Points to the latest public version of the package. If none are public, points to an arbitrary version with `is_active=True` instead

Also just spelling this out made me wonder: in which scenario would `self.available_versions.first()` return `None` but still have a version with `is_active=True`? Shouldn't the former also rule out the latter?

Reply:
is_active isn't the only way we determine visibility (and thusly availability). Right now, all versions don't have visibility so they're technically public, and a "public version of the package" might not be available because of listing review status anyway, so we're actually staying consistent to that behavior by having the same hold true for a rejected version.

Reply:
No but the query `available_versions` uses is only filtering for `is_active=True`, is it not? So what does the added if-clause here actually even do? This doesn't even touch package listings & their review status at all

Reply:
I'm still thinking about this, but I know for certain the _listing_ review status should have nothing to do with how the latest field is cached here as it's an entirely different use case (listing status is community-bound, this is not).

What might be of concern is the newly added version review status, but conceptually that should be analogous to how the current `is_active` field already works, so I suspect my previous question is still valid: in which scenario should the `available_versions` result set be different from the filter being applied in the added if-clause here, outside of ordering? I can't immediately think of any such scenario

Reply:
available_versions also filters by visibility status. Having .latest fall back to a non-visible but still active version means we don't need to change all the cached properties and views we're already using. I think this slight change in behavior to .latest is far less complicated than having an available and unavailable property for everything, or directly getting the first active version everywhere we need one of those properties.

Reply:
> available_versions also filters by visibility status.

Yes but my question is; in which scenario should the way the `is_active` flag has been used in the past not be aligned with the visibility status in the future? Conceptually they are the same, the visibility flag system just has more versatility. 

Really the `is_active` flag should not even be directly queried against anywhere but rather it should only be accounted for in the visibility query directly, I just don't expect all of the queries to be updated in this PR to reflect that yet.

Phrasing the above differently, the visibility flags are a replacement for what `is_active` has been used for in the past, so conceptually the two should be considered the same as far as this conversation goes.

Reply:
Would it make sense to replace `self.latest = self.versions.filter(is_active=True).first()` with `self.latest = self.unavailable_versions.first()` then?

Reply:
Why not omit any & all filtering entirely at that point?

Either way I'd still very much argue that field should only ever contain the latest published version or be left empty just for simplicity's sake. If it really would be fine to have unpublished versions saved there, why would we care about filtering it at all?

Reply:
I've thought about this quite a bit now and can't figure out a way that would not make this change a bad idea in some shape or form. The only reason it would "work" now is because there's implicit behavior elsewhere in the project that happens to produce a sort of "toggle state" between when the field is supposed to only ever point to a public version and when it's allowed to point to an unpublished version, and I can't guarantee that to even be 100% correct without investing significantly more time into reviewing all the possible scenarios & edge cases that could turn up. Even then I can't say I'd be confident I didn't miss something.

To keep things simple, let's just acknowledge that at the very least the majority of the time it would be incorrect to have an unavailable version stored in the field, which still leaves me wondering what's the benefit of this change? Surely it would be less costly in terms of complexity to simply not render the contents on the package detail page if the latest version is empty?

Reply:
I addressed this with https://github.com/thunderstore-io/Thunderstore/pull/1073/commits/c3a2890833fbfaf307f8a7c7235331cbd6e18501 and https://github.com/thunderstore-io/python-packages/pull/22/commits/83484757efa3002999f8c8b926ef39d575c1eb3a

There still seem to be serializers / APIs that don't expect latest to ever be None, but I don''t know if it'll ever cause issues in reality.